docs/ares_parse_txt_reply.3
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\"
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies and that both that copyright
.\" notice and this permission notice appear in supporting
.\" documentation, and that the name of M.I.T. not be used in
.\" advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\" M.I.T. makes no representations about the suitability of
.\" this software for any purpose.  It is provided "as is"
.\" without express or implied warranty.
.\"
.\" SPDX-License-Identifier: MIT
.\"
.TH ARES_PARSE_TXT_REPLY 3 "27 October 2009"
.SH NAME
ares_parse_txt_reply \- Parse a reply to a DNS query of type TXT
.SH SYNOPSIS
.nf
#include <ares.h>

int ares_parse_txt_reply(const unsigned char* \fIabuf\fP, int \fIalen\fP,
                         struct ares_txt_reply **\fItxt_out\fP);

int ares_parse_txt_reply_ext(const unsigned char* \fIabuf\fP, int \fIalen\fP,
                             struct ares_txt_ext **\fItxt_out\fP);
.fi
.SH DESCRIPTION
The \fIares_parse_txt_reply(3)\fP function parses the response to a query of
type TXT into a linked list (one element per sub-string) of
.IR "struct ares_txt_reply" " (" "struct ares_txt_ext" ")"
The parameters
.I abuf
and
.I alen
give the contents of the response.  The result is stored in allocated
memory and a pointer to it stored into the variable pointed to by
.IR txt_out .
It is the caller's responsibility to free the resulting
.IR txt_out
structure when it is no longer needed using the function
\fBares_free_data(3)\fP.

The structure 
.I ares_txt_reply
contains the following fields:
.nf
struct ares_txt_reply {
  struct ares_txt_reply  *next;
  unsigned int  length;
  unsigned char *txt;
};
.fi

The structure
.I ares_txt_ext
contains the following fields:
.nf
struct ares_txt_ext {
  struct ares_txt_ext  *next;
  unsigned int  length;
  unsigned char *txt;
  unsigned char record_start;
};
.fi
The
.I record_start
field in
.I struct ares_txt_ext
is 1 if this structure is a start of a TXT record, and 0 if the structure is a
continuation of a previous record. The linked list of the
.I struct ares_txt_ext
will have at least one item with
.I record_start
equal to 1, and may have some items with
.I record_start
equal to 0 between them.

These sequences of
.I struct ares_txt_ext
(starting from the item with
.I record_start
equal to 1, and ending right before the record start item) may be treated as
either components of a single TXT record or as a multi-parted TXT record,
depending on particular use case.
.SH RETURN VALUES
.BR "ares_parse_txt_reply" " (" "ares_parse_txt_reply_ext" ")"
can return any of the following values:
.TP 15
.B ARES_SUCCESS
The response was successfully parsed.
.TP 15
.B ARES_EBADRESP
The response was malformatted.
.TP 15
.B ARES_ENODATA
The response did not contain an answer to the query.
.TP 15
.B ARES_ENOMEM
Memory was exhausted.
.SH AVAILABILITY
This function was first introduced in c-ares version 1.7.0.
.SH SEE ALSO
.BR ares_query (3)
.BR ares_free_data (3)
.SH AUTHOR
Written by Jakub Hrozek <jhrozek@redhat.com>, on behalf of Red Hat, Inc http://www.redhat.com
.PP
Amended by Fedor Indutny <fedor@indutny.com>, on behalf of PayPal, Inc https://www.paypal.com


src/lib/ares_parse_txt_reply.c
/* MIT License
 *
 * Copyright (c) 1998 Massachusetts Institute of Technology
 * Copyright (c) 2009 Jakub Hrozek
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */

#include "ares_setup.h"

#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif
#ifdef HAVE_NETDB_H
#  include <netdb.h>
#endif
#ifdef HAVE_ARPA_INET_H
#  include <arpa/inet.h>
#endif

#include "ares_nameser.h"

#ifdef HAVE_STRINGS_H
#  include <strings.h>
#endif

#include "ares.h"
#include "ares_dns.h"
#include "ares_data.h"
#include "ares_private.h"

static int ares__parse_txt_reply(const unsigned char *abuf, size_t alen,
                                 ares_bool_t ex, void **txt_out)
{
  size_t               substr_len;
  size_t               qdcount, ancount, i;
  const unsigned char *aptr;
  const unsigned char *strptr;
  ares_status_t        status;
  int                  rr_type, rr_class;
  size_t               rr_len;
  size_t               len;
  char                *hostname = NULL, *rr_name = NULL;
  struct ares_txt_ext *txt_head = NULL;
  struct ares_txt_ext *txt_last = NULL;
  struct ares_txt_ext *txt_curr;

  /* Set *txt_out to NULL for all failure cases. */
  *txt_out = NULL;

  /* Give up if abuf doesn't have room for a header. */
  if (alen < HFIXEDSZ) {
    return ARES_EBADRESP;
  }

  /* Fetch the question and answer count from the header. */
  qdcount = DNS_HEADER_QDCOUNT(abuf);
  ancount = DNS_HEADER_ANCOUNT(abuf);
  if (qdcount != 1) {
    return ARES_EBADRESP;
  }
  if (ancount == 0) {
    return ARES_ENODATA;
  }

  /* Expand the name from the question, and skip past the question. */
  aptr   = abuf + HFIXEDSZ;
  status = ares__expand_name_for_response(aptr, abuf, alen, &hostname, &len,
                                          ARES_TRUE);
  if (status != ARES_SUCCESS) {
    return (int)status;
  }

  if (aptr + len + QFIXEDSZ > abuf + alen) {
    ares_free(hostname);
    return ARES_EBADRESP;
  }
  aptr += len + QFIXEDSZ;

  /* Examine each answer resource record (RR) in turn. */
  for (i = 0; i < ancount; i++) {
    /* Decode the RR up to the data field. */
    status = ares__expand_name_for_response(aptr, abuf, alen, &rr_name, &len,
                                            ARES_FALSE);
    if (status != ARES_SUCCESS) {
      break;
    }
    aptr += len;
    if (aptr + RRFIXEDSZ > abuf + alen) {
      status = ARES_EBADRESP;
      break;
    }
    rr_type   = DNS_RR_TYPE(aptr);
    rr_class  = DNS_RR_CLASS(aptr);
    rr_len    = DNS_RR_LEN(aptr);
    aptr     += RRFIXEDSZ;
    if (aptr + rr_len > abuf + alen) {
      status = ARES_EBADRESP;
      break;
    }

    /* Check if we are really looking at a TXT record */
    if ((rr_class == C_IN || rr_class == C_CHAOS) && rr_type == T_TXT) {
      /*
       * There may be multiple substrings in a single TXT record. Each
       * substring may be up to 255 characters in length, with a
       * "length byte" indicating the size of the substring payload.
       * RDATA contains both the length-bytes and payloads of all
       * substrings contained therein.
       */

      strptr = aptr;
      while (strptr < (aptr + rr_len)) {
        substr_len = (unsigned char)*strptr;
        if (strptr + substr_len + 1 > aptr + rr_len) {
          status = ARES_EBADRESP;
          break;
        }

        /* Allocate storage for this TXT answer appending it to the list */
        txt_curr = ares_malloc_data(ex ? ARES_DATATYPE_TXT_EXT
                                       : ARES_DATATYPE_TXT_REPLY);
        if (!txt_curr) {
          status = ARES_ENOMEM;
          break;
        }
        if (txt_last) {
          txt_last->next = txt_curr;
        } else {
          txt_head = txt_curr;
        }
        txt_last = txt_curr;

        if (ex) {
          txt_curr->record_start = (strptr == aptr);
        }
        txt_curr->length = substr_len;
        txt_curr->txt = ares_malloc(substr_len + 1 /* Including null byte */);
        if (txt_curr->txt == NULL) {
          status = ARES_ENOMEM;
          break;
        }

        ++strptr;
        memcpy((char *)txt_curr->txt, strptr, substr_len);

        /* Make sure we NULL-terminate */
        txt_curr->txt[substr_len] = 0;

        strptr += substr_len;
      }
    }

    /* Propagate any failures */
    if (status != ARES_SUCCESS) {
      break;
    }

    /* Don't lose memory in the next iteration */
    ares_free(rr_name);
    rr_name = NULL;

    /* Move on to the next record */
    aptr += rr_len;
  }

  if (hostname) {
    ares_free(hostname);
  }
  if (rr_name) {
    ares_free(rr_name);
  }

  /* clean up on error */
  if (status != ARES_SUCCESS) {
    if (txt_head) {
      ares_free_data(txt_head);
    }
    return (int)status;
  }

  /* everything looks fine, return the data */
  *txt_out = txt_head;

  return ARES_SUCCESS;
}

int ares_parse_txt_reply(const unsigned char *abuf, int alen,
                         struct ares_txt_reply **txt_out)
{
  if (alen < 0) {
    return ARES_EBADRESP;
  }
  return ares__parse_txt_reply(abuf, (size_t)alen, ARES_FALSE,
                               (void **)txt_out);
}

int ares_parse_txt_reply_ext(const unsigned char *abuf, int alen,
                             struct ares_txt_ext **txt_out)
{
  if (alen < 0) {
    return ARES_EBADRESP;
  }
  return ares__parse_txt_reply(abuf, (size_t)alen, ARES_TRUE, (void **)txt_out);
}
